packages/create-plugin/src/utils/utils.packageManager.ts

```diff
@@ -0,0 +1,105 @@
+const PNPM_LOCKFILE = 'pnpm-lock.yaml';
+const YARN_LOCKFILE = 'yarn.lock';
+const SUPPORTED_PACKAGE_MANAGERS = ['npm', 'yarn', 'pnpm'];
+const LEGACY_PACKAGE_MANAGER = { packageManagerName: 'yarn', packageManagerVersion: '1.22.19' };
+
+export type PackageManager = {
+  packageManagerName: string;
+  packageManagerVersion: string;
+};
+
+export function getPackageManagerFromUserAgent(): PackageManager {
+  const agent = process.env.npm_config_user_agent;
+
+  const [name, versionWithText] = agent.split('/');
+  const [version] = versionWithText.split(' ');
+
+  if (SUPPORTED_PACKAGE_MANAGERS.includes(name)) {
+    return { packageManagerName: name, packageManagerVersion: version.trimEnd() };
+  }
+
+  return LEGACY_PACKAGE_MANAGER;
+}
+
+export function getPackageManagerWithFallback() {
+  const packageManagerFromPackageJson = getPackageManagerFromPackageJson();
+  if (packageManagerFromPackageJson) {
+    return packageManagerFromPackageJson;
+  }
+
+  const packageManagerFromLockFile = getPackageManagerFromLockFile();
+  if (packageManagerFromLockFile) {
+    return packageManagerFromLockFile;
+  }
+
+  return LEGACY_PACKAGE_MANAGER;
+}
+
+export function getPackageManagerInstallCmd(packageManagerName: string) {
+  switch (packageManagerName) {
+    case 'yarn':
+      return 'yarn install --immutable --prefer-offline';
+
+    case 'pnpm':
+      return 'pnpm install --frozen-lockfile --prefer-offline';
+
+    default:
+      return 'npm ci';
+  }
+}
+
+function getPackageManagerFromLockFile(): PackageManager | undefined {
+  const closestLockfilePath = findUpSync([YARN_LOCKFILE, PNPM_LOCKFILE, NPM_LOCKFILE]);
+  if (!Boolean(closestLockfilePath)) {
+    return undefined;
+  }
+
+  const closestLockfile = closestLockfilePath && basename(closestLockfilePath);
+
+  try {
+    if (closestLockfile === NPM_LOCKFILE) {
+      const npmVersionCommand = spawnSync('npm', ['--version'], {
+        shell: true,
+      });
+      const packageManagerVersion = npmVersionCommand.stdout.toString().trimEnd();
+      return { packageManagerName: 'npm', packageManagerVersion };
+    }
+
+    if (closestLockfile === PNPM_LOCKFILE) {
+      const pnpmVersionCommand = spawnSync('pnpm', ['--version'], {
+        shell: true,
+      });
+      const packageManagerVersion = pnpmVersionCommand.stdout.toString().trimEnd();
+      return { packageManagerName: 'pnpm', packageManagerVersion };
+    }
+
+    if (closestLockfile === YARN_LOCKFILE) {
+      const yarnVersionCommand = spawnSync('yarn', ['--version'], {
+        shell: true,
+      });
+      const packageManagerVersion = yarnVersionCommand.stdout.toString().trimEnd();
+      return { packageManagerName: 'yarn', packageManagerVersion };
+    }
+
+    return undefined;
+  } catch (error) {
+    console.error('Failed to find package manager from lock file. Have you installed dependencies?');
+    throw Error(error);
+  }
+}
+
+function getPackageManagerFromPackageJson(): PackageManager | undefined {
```

Comment:
my paraonid mind tells me that we can't really trust the package.json to contain the correct package manager and specially since this is also an experimental feature https://nodejs.org/docs/latest-v18.x/api/all.html#all_packages_packagemanager

my suggestion is to rely on lock files and fallback to this.

Reply:
Hmm, I am not sure, as we are persisting this field in the package.json during scaffolding / updating / migrating I would say we can rely on it. As we are using the same signature (`"<package manager name>@<version>"`) as Node.js and Corepack, I don't think this should conflict (we only override this property if it was not set before)?

Reply:
As @leventebalogh mentions we're setting this ourselves based on either the lock files or the npm_user_agent if it doesn't already exist. Whilst corepack is experimental this package.json property is already widely used in projects (courtesy of yarn berry and pnpm) which gives me the confidence to use it here too. It's also being used by the pnpm github action to know which version of pnpm to install when running the workflows.